CARDS-1705: Create CSV to Vocabulary import script

acrowthe opened the pull request:
Create import script for converting .csv files to .obo files
- Update Cards4Kids vocabularies to use these obo files instead of the
  old owl files

Testing notes: 
Make sure that the Cards4Kids vocabularies in the admin dashboard match the vocabularies as defined in the IMPACT spreadsheet. Make sure that the IMPACT vocabulary questions only offer the values in the name or details column of the vocabulary spreadsheet as options.

Results of the import script are the .obo files, which are imported automatically when the Cards4Kids run mode is used. Starting up Cards4Kids for the first time will take longer than normal as the vocabularies are imported (a few minutes). The DeviceMasterlist in particular takes a while to import due to the large amount of terms in it.

![image](https://user-images.githubusercontent.com/20056751/161161015-2845a474-d5b8-44e8-8a87-5b02a262a5b5.png)

acrowthe committed:
CARDS-1286: Cards4Kids Import
- NPC, STS, PAC3 and PC4 forms current unable to load due to broken conditionals
- Remaining conditionals in progress due to refactor to handle strings in quotations

acrowthe committed:
CARDS-1286: Cards4Kids Import
- All forms ready

acrowthe committed:
CARDS-1286: Cards4Kids Import
- Fix rebase conflicts due to lfs > cards rename

acrowthe committed:
CARDS-1286: Cards4Kids Import
- Parse units from descriptions
- Escape markdown charaters
- Remove descriptions that were identical to question titles or expressions

acrowthe committed:
CARDS-1286: Import Cards4Kids Questionnaires
Fix NPC-QIC description parsing
- Now correctly parses min and max values from this form
- No longer parses incorrect parts of the description into conditionals

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Draft data modeling commit

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Corrected data modeling

Commit:
[misc] Removed unused import

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
WIP: Form rendering, Subject, Subject timeline, form to text adapter

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
WIP: Questionnaire wizard

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
WIP: Some bugfix

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
WIP: Form rendering bugfixes

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
WIP: some UI polishing

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
WIP: QuestionMatrix Commit editor

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Subject view rendering

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Removed QuestionMatrix data type

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Some more adjustments regarding removing QuestionMatrix data type

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Renamed component to QuestionMatrix

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Checking props existing before copying to the question

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Added missing props setting when option is copied to the new question

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Fixed answer instructions alignment;
Do not show questions without answers in form view mode

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Addressed codereview comments:
set displayMode=list on sub-questions
reverted unnecessary changes
refactored QuestionMatrix to avoid changes to Question.jsx

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Fixed show section in view mode if mandatory

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Fixed Quick Search NPE while searching for one of the AnswerOption values associated with a Question matrix results

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Moved highlight logic to the Question to avoid code duplication

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Addressed codereview comments: optimisation and refactoring

marta- committed:
CARDS-1400: New Questionnaire entry type: Question matrix
Added test questionnaire

marta- committed:
CARDS-1400: New Questionnaire entry type: Question matrix
Bug fix: for question matrices with numeric datatype, the selection appears to be missing on re-edit

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Copy all matrix sextion props to new Question child node

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Corrected comment

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Renamed gridClasses to cardClasses

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Code cleanup

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Fixed 409 repository state conflicting with request when saving newly created Form first time with matrix quetion with an answer by clicking save and then save&view

sdumitriu committed:
CARDS-1400: New Questionnaire entry type: Question matrix
Fixed AnswerOption copying on initial content import

sdumitriu committed:
CARDS-1400: New Questionnaire entry type: Question matrix
Fixed wrong property type for jcr:primaryType on copies answers

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Minor UI corrections

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Minor UI corrections

Commit:
CARDS-1070: Changing the question variable name erases the previously entered AnswerOptions

acrowthe committed:
CARDS-1286: Cards4Kids Import
- Implement matrix questions

acrowthe committed:
CARDS-1286: Cards4Kids Import
Fix date computed questions
- Implement new entryMode computed format
- Fix issue entering dates
- Fix error parsing <= and >=

acrowthe committed:
CARDS-1286: Cards4Kids Import
Add pagination

acrowthe committed:
CARDS-1286: Cards4Kids Import
Address review comments
- General ACTIONS improvement
- NPC-QIC required improvement
- cnoc 6_21 option title clean up

acrowthe committed:
CARDS-1400: New Questionnaire entry type: Question matrix
- Implement pageActive support

acrowthe committed:
CARDS-1400: New Questionnaire entry type: Question matrix
- Remove unneeded pageActive from administration page

acrowthe committed:
CARDS-1286: Cards4Kids Import
- Add in vocabularies for IMPACT form
- Address review comments on ACTION and IMPACT

acrowthe committed:
CARDS-1705: Create CSV to Vocabulary import script
Create import script for converting .csv files to .obo files
- Update Cards4Kids vocabularies to use these obo files instead of the
  old owl files

acrowthe committed:
CARDS-1705: Create CSV to Vocabulary import script
Fix non-functional vocabularyFilters on IMPACT vocabulary questions

sashaandjic commented:
Can this be changed? To replace Value with actual device type
![image](https://user-images.githubusercontent.com/23243244/161403050-29b00ae3-340b-4391-b6ff-6dccc21d7108.png)
![image](https://user-images.githubusercontent.com/23243244/161403061-6018bbb3-7ecf-4144-82db-00d727dc9c56.png)

acrowthe commented:
> Can this be changed? To replace Value with actual device type

3 options:
1. Get rid of Value. The reason Value is in there is so that the vocabulary fields can be filtered to "Value", so you'll be able to select Categories in Device Type fields.
2. Try replacing is_a value with subsetdef. I haven't looked into subsetdef, so this may keep the filter functionality without being listed as a type
3. Try reordering it to read
> Ablation Cather
> St. Jude Medical Center
> Value

I'll give 2 a try on Monday, but if that doesn't work I'm not sure which is best from 1 and 3

sashaandjic commented:
I should not be able to select Value
![image](https://user-images.githubusercontent.com/23243244/161576568-05074a98-3cd0-4e93-8e9a-65fd1936ee70.png)
![image](https://user-images.githubusercontent.com/23243244/161576594-874cc1c3-1c56-48db-bda6-37a16bc66af8.png)

acrowthe commented:
Changes to make per standup discussion:
- Remove `Value` and `Category` terms, as well as `Value` filter on vocabulary questions
- Add all subcategories to the `comments` field on `Terms` so they are searchable
- Check how many duplicate labels there are. If there are a substantial amount, append subcategories to the label to remove duplicates
- Set the ID of all category terms to be a unique negative number

acrowthe committed:
CARDS-1705: Create CSV to Vocabulary import script
General ontology structure improvements
- Remove `Value` and `Category terms
- Change category term IDs to follow the format `_#`
- Add category names to term descriptions for improved searching
- Improve duplicate term naming

acrowthe commented:
Removed `Value` and `Category` Terms:
![image](https://user-images.githubusercontent.com/20056751/162070691-07e54907-fcd5-4c2c-beea-2a3f6423f97a.png)

Added more details to the names of duplicated terms:
![image](https://user-images.githubusercontent.com/20056751/162070350-533def00-2c14-44df-99af-3bcf013b9b4c.png)

Added subcategories to the descriptions:
![image](https://user-images.githubusercontent.com/20056751/162070431-cf3c4b97-3c3a-45bc-b669-1f664fd07a0d.png)

Added a prefix `_` to category IDs. Ideally this would be a `-` so we could easily filter negative vs. positive IDs when required. However, `-` are not currently allowed in ontology IDs, and I haven't found a list of valid ontology ID characters that say they should be allowable.
![image](https://user-images.githubusercontent.com/20056751/162070751-62f93d37-80cc-4024-a1da-1f3460483656.png)

https://github.com/data-team-uhn/cards/blob/a576258b5087461d16b483aeca30a8d0b2de3994/modules/vocabularies/src/main/java/io/uhndata/cards/vocabularies/internal/OntologyIndexerUtils.java#L74

sashaandjic reviewed (approved):
tested - works as we discussed, however I'm not happy with possibility of selecting an invalid item (e.g. subcategory like device type or device manufacturer))

acrowthe commented:
> I'm not happy with possibility of selecting an invalid item
@sashaandjic 

For removing the ability to select invalid terms, there are 3 options I see:
1. Re-add value (or a renamed equivalent) as a term, as it was before
2. Re-add value as a subset. The UI would be the same as it is currently, but would require editing the vocabulary filters to support subsets
3.  Convert the category IDs to be negative numbers and exclude negative numbers. Would require:
  - Editing the vocabulary importers to support `-` in IDs, assuming that is allowed by OBO standards
  - Updating vocabulary questions to support min/max values in cases where IDs are numeric
  - Updating vocabulary filters to support min/max values in cases where IDs are numeric
  
 I think 3 is preferable if IDs are able to include dashes

sashaandjic commented:
I the worst case scenario, we can dich the vocabulary field and offer users `BROWSE` vocabulary button next to the numerical filed that has range defined as specified in the spredsheet. This would also be a huge improvement for the current workflow.
@marta- WDYT?

acrowthe committed:
CARDS-1705: Create CSV to Vocabulary import script
- Convert category IDs to be negative, so they are ready for when vocabulary question ranges are implemented

acrowthe commented:
Category IDs are now negative while regular terms are positive numbers - When vocabulary questions get improved validation, it should be easy to prevent selection of categories now

sashaandjic commented:
Can we remove `ID` from `Device ID`?
![image](https://user-images.githubusercontent.com/23243244/162562291-13059a9f-e202-48ab-85de-9562e45173f8.png)

sashaandjic commented:
> Category IDs are now negative while regular terms are positive numbers - When vocabulary questions get improved validation, it should be easy to prevent selection of categories now

[CARDS-1721](https://phenotips.atlassian.net/browse/CARDS-1721)

marta- reviewed (approved):
@acrowthe I think this looks good and is OK to merge into 1286 rather than leaving it on a separate branch. If any issues are detected we can work on refining it alongside with the actual questionnaires.
